ckanext/activity/tests/conftest.py

```diff
@@ -13,3 +20,24 @@ class ActivityFactory(CKANFactory):
     class Meta:
         model = Activity
         action = "activity_create"
+
+
+@pytest.fixture(autouse=True, scope="module")
+def apply_activity_migrations():
+    plugin = "activity"
+
+    factories.fake.unique.clear()
+    helpers.reset_db()
+
+    if not ckan.plugins.plugin_loaded(plugin):
+        ckan.plugins.load(plugin)
+
+    _run_migrations(plugin, version="head", forward=True)
```

Comment:
running migrations as a side effect of importing a module, even if only for tests, is very surprising behavior. Can't you have the migrations run when a test that needs them is run instead?

Reply:
@wardi I changed it to only running the migrations in test classes that need them. Is it okay now?

Reply:
The most pytest-like approach here is to define [session-scoped reset_db_once and function scoped clean_db](https://github.com/ckan/ckanext-editable-config/blob/master/ckanext/editable_config/tests/conftest.py#L23-L32) fixtures instead of creating a new autouse-fixture that does such low-level things.

In pytest philosophy, the fixture should always do whatever makes sense for the current situation. for example, in CKAN core tests `clean_db` ensures that all the tables that CKAN uses are created and empty.

If we use `clean_db` during activity tests, it should as well guarantee that **all tables are created and empty**. But because we are inside a plugin that defines additional tables(or migrations), the definition of "all tables" is a bit different. Not it's all the tables that are affected by CKAN core migrations + all the tables that are affected by `activity` migrations. That's why I suggest registering activity-specific `clean_db`:
```python
@pytest.fixture()
def clean_db(reset_db, migrate_db_for):
    reset_db()
    migrate_db_for("activity")
```

Do not load `activity` plugin if it's missing. Fixtures should not be too smart. It's up to the test to include the required plugin alongside with `with_plugin` fixture. I.e, a test that interacts with DB inside the activity plugin will look like this:
```python
@pytest.mark.ckan_config("ckan.plugins", "activity")
@pytest.mark.usefixtures("with_plugins", "clean_db")
class TestActivity: ...
```
You may think that it's the wrong approach. What if the author of the test includes `clean_db` and forgets to enable the `activity` plugin? As you may have guessed, it will cause an error. But it's an expected and obvious outcome. If you interact with DB but forgot to include `clean_db`, you also can get an error. If you call datastore action but forgot to load the corresponding plugin - you'll get an error. Sometimes test authors **have to** understand the structure and do explicitly extra work. Fixtures should simplify our life, but they should never do anything that has no relation to their names.

Of course, you can load the plugin, as you do at the moment. But it creates more problems:
* Should this plugin be unloaded after the test? Even if it was enabled before? Wouldn't it be better to leave environment in the exactly the same state it was before?
* How can we test behavior when someone runs migrations without the plugin if fixture that runs migrations loads the plugin unconditionally?


`reset_db_once` can be defined in order to modify `non_clean_db`. You can redefine `non_clean_db` directly, but it's more complex, so overriding `reset_db_once` is just a simple workaround

Reply:
@smotornyuk  Thanks for the explanation.

The `clean_db` tests work well now, but the `non_clean_db` ones fail with `Plugin 'activity' cannot be loaded.` Any idea why? The same error was showing with `clean_db`, but was fixed after I changed the order of `clean_db` and `with_plugins` in the `usefixtures` mark, it's not working for `non_clean_db` though.

Also some tests from other modules (ex. `ckan/tests/lib/dictization/test_dictization.py`) fail because of the missing activity migration, but they don't activate the activity plugin. I was wondering what to do with those?

Reply:
@smotornyuk Thanks for the explanation.

The clean_db tests work well now, but the non_clean_db ones fail with Plugin 'activity' cannot be loaded. Any idea why? The same error was showing with clean_db, but was fixed after I changed the order of clean_db and with_plugins in the usefixtures mark, it's not working for non_clean_db though.

Also, some tests from other modules (ex. ckan/tests/lib/dictization/test_dictization.py) fail because of the missing activity migration, but they don't activate the activity plugin. I was wondering what to do with those?

Reply:
@smotornyuk Thanks for the explanation.

The `clean_db` tests work well now, but the `non_clean_db` ones fail with `Plugin 'activity' cannot be loaded.` Any idea why? The same error was showing with `clean_db`, but was fixed after I changed the order of `clean_db` and `with_plugins` in the `usefixtures` mark, it's not working for `non_clean_db` though.

Also, some tests from other modules (ex. `ckan/tests/lib/dictization/test_dictization.py`) fail because of the missing activity migration, but they don't activate the activity plugin. I was wondering what to do with those?

Reply:
@smotornyuk Thanks for the explanation.

The `clean_db` tests work well now, but the `non_clean_db` ones fail with `Plugin 'activity' cannot be loaded.` Any idea why? The same error was showing with `clean_db`, but was fixed after I changed the order of `clean_db` and `with_plugins` in the `usefixtures` mark, it's not working for `non_clean_db` though.

Reply:
My bad, I forgot about the order of fixture loading. In this scenario we cannot use `non_clean_db`, so switch to `clean_db` everywhere inside the `activity` plugin.

Explanation, if you are interested:

Every pytest fixture has a scope - it's the frequency of calling the fixture. By default, every fixture is called once for every separate test(`function` scope), but `reset_db_once` has a `session` scope instead, so it's called only once during the whole test session and, as a result, cleans DB only once, no matter how many times it's used.

In the initialization stage, pytest sorts fixtures by scope first and then by the order they are listed in the test. For example, in the following test:
```python
def test_smth(FUNCTION_x, SESSION_y, FUNCTION_z):
   ...
```
assuming the prefix reflects the scope of a fixture, pytest will call them in the following order:
* SESSION_y - because it has the highest scope
* FUNCTION_x - because among all fixtures with function scope, it's listed first in test arguments
* FUNCTION_z

This means that `reset_db_once` is always initialized first, before the function-scoped `with_plugins`. You cannot change the scope of `reset_db_once` to the `function`, because it will be called for every test then and essentially will turn into simple `clean_db`. Neither you can include `with_plugins` as a dependency of `reset_db_once`, because you are not allowed to include narrow-scoped fixtures into wide-scoped ones. 

So the only sensible option is to use `clean_db` instead of `non_clean_db` here. There are only a few tests that are using this fixture, so it won't create any visible downgrade in the execution time of tests